### PR TITLE
Fix #1869 cleanup ibisstore until there are no more records

### LINKED ISSUE #1869: Job ibisstore
**Describe the issue**
(Sorry for writing in Dutch, I first wrote this in an email to Jaco. If you need the English version please contact me)

In het nieuwe Frank!Framework is de cleanupDatabase job iets anders dan tov het oude framwork:

DELETE FROM IBISSTORE WHERE MESSAGEKEY IN (SELECT TOP 25000 MESSAGEKEY FROM IBISSTORE WITH (readpast) WHERE TYPE IN ('L','A') AND EXPIRYDATE < [datumvanvandaag])

In de nieuwe versie wordt een SELECT TOP 25000 gedaan, dit voorkomt een timeout in het geval dat er hondderdduizenden berichten zijn.
(De oude versie gaf ook problemen gezien er een timeout ontstond als er te veel berichten waren.)

Maar: in het geval van ibis4emailhandler komen er dagelijks zo’n 50 a 60 duizend berichten bij in de tabel Ibisstore. Kortom: maar de helft wordt opgeruimd en dit veroorzaakte laatst een incident. Ik ben al met de dba'er bezig om dit op te lossen en om te kijken of er aan de db kant nog een extra handigheidje kan worden ingebouwd.

Hebben jullie hier misschien nog goede ideeën over? Of is het bijv. een mogelijkheid om de scheduler (voor de query SELECT TOP 25000…) te laten loop-en zolang er meer dan 25000 berichten zijn die aan de zoekopdracht voldoen?

**Reporter**
Laura Nauta-Timp

### PR DESCRIPTION
**Pull request by alisihab**
Closes #1869

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java [hunk 1]
```diff
@@ -755,9 +756,18 @@ private void cleanupDatabase(IbisManager ibisManager) {
 					+ "') AND " + mlo.getExpiryDateField() + " < ?"+(maxRows>0?" FETCH FIRST "+maxRows+ " ROWS ONLY":"")+")");
 					qs.setSqlDialect(Dbms.ORACLE.getKey());
 				}
-
-				Message result = qs.sendMessage(new Message(query), null);
-				log.info("result [" + result + "]");
+				boolean fetchedAllRecords = false;
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java:759]: I would call this 'deletedAllRecords'

File: core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java [hunk 2]
```diff
@@ -755,9 +756,18 @@ private void cleanupDatabase(IbisManager ibisManager) {
...
+				while(!fetchedAllRecords) {
+					Message result = qs.sendMessage(new Message(query), null);
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java:761]: You need to execute each call to sendMessage in a separate transaction

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java [hunk 3]
```diff
@@ -755,9 +756,18 @@ private void cleanupDatabase(IbisManager ibisManager) {
...
+					String resultString = result.asString();
+					log.info("deleted [" + resultString + "] rows");
+					int numberOfRowsAffected = Integer.valueOf(resultString);
+					if(numberOfRowsAffected<maxRows) {
+						fetchedAllRecords = true;
+					} else {
+						log.info("executing the query again!");
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java:768]: for job [cleanupDatabase]

File: core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java [hunk 4]
```diff
@@ -1209,8 +1206,7 @@ public String getJmsRealm() {
 	}
 
 	@IbisDoc({"Optional Locker, to avoid parallel execution of the Job by multiple threads or servers. The Job is NOT executed when the lock cannot be obtained, " +
-				"e.g. in case another thread, may be in another server, holds the lock and does not release it in a timely manner. " +
-				"N.B. To retain the lock, even after the transaction of the Job rolled back, set Lockers transactionAttribute=\"RequiresNew\" and type=\"P\"; Otherwise, the Locker will join the transaction of the Job and/or the Lock will be released."})
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java:1213]: Is this no longer true?

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java (hunk 4 above)
Comment (alisihab) [reply core/src/main/java/nl/nn/adapterframework/scheduler/JobDef.java:1213]: Job execution is no longer transactional. There is no transaction for locker's transaction to join

**Review by nielsm5 — APPROVED** [review]

**Review by gvanbrakel — APPROVED** [review]

**Comment by nielsm5** [comment]
TxManager has been moved to the locker in pr #1387

**Pull request merged**